src/librustc_mir/transform/inline.rs

```diff
@@ -701,14 +695,14 @@ impl<'a, 'tcx> MutVisitor<'tcx> for Integrator<'a, 'tcx> {
             },
             Place {
                 base: PlaceBase::Static(box Static {
-                    kind: StaticKind::Promoted(promoted),
+                    kind: StaticKind::Promoted(_, substs),
                     ..
                 }),
                 projection: None,
             } => {
-                if let Some(p) = self.promoted_map.get(*promoted).cloned() {
-                    *promoted = p;
-                }
+                let adjusted_substs = substs.subst(self.tcx, self.callsite.substs);
```

Comment:
This should have already happened, it will ICE here I think

Reply:
Hmm... I don't see anywhere else in the inliner that handles promoted MIR. The following also seems to work correctly:

```
narada:temp wesley$ cat test.rs
fn main() {
    let _ = make_empty::<usize>();
}

#[inline]
fn make_empty<T>() -> &'static [T] {
    &[]
}

narada:temp wesley$ RUST_BACKTRACE=1 rustc -Z mir-opt-level=3 --emit=mir test.rs 
```

Reply:
This isn't about promoted MIR, this is about the index into a promoted's mir::Body and the relevant substs needed for it. Since this is part of the current function's `mir::Body`, it will get substituted in https://github.com/rust-lang/rust/blob/13d68595b1c9818a2f5dac1bcf5fa6c804a5f2af/src/librustc_mir/transform/inline.rs#L121

Reply:
you should just remove this arm entirely now

Reply:
So, if I remove the arm, I get this:

```
error: internal compiler error: src/librustc/ty/subst.rs:557: type parameter `T/#0` (T/0) out of range when substituting (root type=Some(T)) substs=[]

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:579:9
stack backtrace:
   0: std::sys_common::backtrace::print
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: rustc::util::common::panic_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: rustc_errors::Handler::span_bug
   7: rustc::util::bug::opt_span_bug_fmt::{{closure}}
   8: rustc::ty::context::tls::with_opt::{{closure}}
   9: rustc::ty::context::tls::with_context_opt
  10: rustc::ty::context::tls::with_opt
  11: rustc::util::bug::opt_span_bug_fmt
  12: rustc::util::bug::span_bug_fmt
  13: <rustc::ty::subst::SubstFolder as rustc::ty::fold::TypeFolder>::fold_ty
  14: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
  15: <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter
  16: rustc::ty::fold::TypeFoldable::fold_with
  17: <rustc_mir::monomorphize::collector::MirNeighborCollector as rustc::mir::visit::Visitor>::visit_place_base
  18: rustc::mir::visit::Visitor::super_rvalue
  19: <rustc_mir::monomorphize::collector::MirNeighborCollector as rustc::mir::visit::Visitor>::visit_rvalue
  20: rustc::mir::visit::Visitor::visit_body
  21: rustc_mir::monomorphize::collector::collect_items_rec
  22: rustc_mir::monomorphize::collector::collect_crate_mono_items::{{closure}}
  23: rustc::util::common::time
  24: rustc_mir::monomorphize::collector::collect_crate_mono_items
  25: rustc::util::common::time
  26: rustc_mir::monomorphize::partitioning::collect_and_partition_mono_items
  27: rustc::ty::query::__query_compute::collect_and_partition_mono_items
  28: rustc::dep_graph::graph::DepGraph::with_task_impl
  29: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  30: rustc_codegen_ssa::base::codegen_crate
  31: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_utils::codegen_backend::CodegenBackend>::codegen_crate
  32: rustc::util::common::time
  33: rustc_interface::passes::start_codegen
  34: rustc::ty::context::tls::enter_global
  35: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  36: rustc_interface::passes::create_global_ctxt::{{closure}}
  37: rustc_interface::passes::BoxedGlobalCtxt::enter
  38: rustc_interface::queries::Query<T>::compute
  39: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::ongoing_codegen
  40: rustc_interface::interface::run_compiler_in_existing_thread_pool
  41: std::thread::local::LocalKey<T>::with
  42: scoped_tls::ScopedKey<T>::set
  43: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
query stack during panic:
#0 [collect_and_partition_mono_items] collect_and_partition_mono_items
end of query stack

```

Reply:
Which makes sense to me because the logging I added shows the replacement is important in this test case:

```
replacing substs [T] with [usize]
```

Reply:
Ok, that means the `subst` tree walker doesn't visit the substs inside the promoted. Please check what walker is used inside the `subst` function and make sure it knows about `StaticKind` and its `SubstsRef`

Reply:
Fixed.